test/python/transpiler/test_remove_final_measurements.py

```diff
@@ -0,0 +1,316 @@
+        the measure destination bit itself is idle.
+        """
+
+        def expected_dag():
+            q0 = QuantumRegister(1, "q0")
+            c0 = ClassicalRegister(2, "c0")
+            qc = QuantumCircuit(q0, c0)
+            qc.x(q0[0]).c_if(c0[0], 0)
+            return circuit_to_dag(qc)
+
+        q0 = QuantumRegister(1, "q0")
+        c0 = ClassicalRegister(2, "c0")
+        qc = QuantumCircuit(q0, c0)
+
+        # make c0[0] busy
+        qc.x(q0[0]).c_if(c0[0], 0)
+
+        # measure into not busy c0[1]
+        qc.measure(0, c0[1])
+
+        dag = circuit_to_dag(qc)
+        dag = RemoveFinalMeasurements().run(dag)
+
+        # c0 should not be removed because it has busy bit c0[0]
+        self.assertListEqual(list(dag.cregs.values()), [c0])
+
+        # note: c0[1] should not be removed even though it is now idle
+        # because it is referenced by creg c0.
+        self.assertListEqual(dag.clbits, list(c0))
+        self.assertEqual(dag, expected_dag())
+
+    def test_overlapping_register_removal(self):
+        """Only registers that become idle directly as a result of
+        final op removal are removed. In this test, a 5-bit creg
+        is implicitly created with its own bits, along with cregs
+        ``c0_upper_3`` and ``c0_lower_3`` which reuse those underlying bits.
+        ``c0_upper_3`` and ``c0_lower_3`` reference only 1 bit in common.
+        A final measure is performed into a bit that exists in ``c0_upper_3``
+        but not in ``c0_lower_3``, and subsequently is removed. Consequently,
+        both ``c0_upper_3`` and the 5-bit register are removed, because they
+        have become unused as a result of the final measure removal.
+        ``c0_lower_3`` remains, because it was idle beforehand, not as a
+        result of the measure removal, along with all of its bits,
+        including the bit shared with ``c0_upper_3``."""
+
+        def expected_dag():
+            q0 = QuantumRegister(3, "q0")
+            c0 = ClassicalRegister(5, "c0")
+            c0_lower_3 = ClassicalRegister(name="c0_lower_3", bits=c0[2:])
+
+            # note c0 is *not* added to circuit!
+            qc = QuantumCircuit(q0, c0_lower_3)
+            return circuit_to_dag(qc)
+
+        q0 = QuantumRegister(3, "q0")
+        c0 = ClassicalRegister(5, "c0")
+        qc = QuantumCircuit(q0, c0)
+
+        c0_upper_3 = ClassicalRegister(name="c0_upper_3", bits=c0[:3])
+        c0_lower_3 = ClassicalRegister(name="c0_lower_3", bits=c0[2:])
+        # Only qc.clbits[2] is shared between the two.
+
+        qc.add_register(c0_upper_3)
+        qc.add_register(c0_lower_3)
+
+        qc.measure(0, c0_upper_3[0])
+
+        dag = circuit_to_dag(qc)
+        dag = RemoveFinalMeasurements().run(dag)
+
+        self.assertListEqual(list(dag.cregs.values()), [c0_lower_3])
+        self.assertListEqual(dag.clbits, list(c0_lower_3))
+        self.assertEqual(dag, expected_dag())
```

Comment:
I understand the test here (`c0` becomes idle, so is removed, whereas `c0_lower_3` was always idle so it stays), and it's the right behaviour to implement, but I wouldn't be surprised if we get a "bug" report about it at some point in the future!